jumanji/environments/logic/graph_coloring/viewer.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Sequence, Tuple
+
+import chex
+import matplotlib.animation as animation
+import matplotlib.cm as cm
+import matplotlib.pyplot as plt
+import numpy as np
+
+import jumanji.environments
+from jumanji.environments.logic.graph_coloring.types import State
+from jumanji.viewer import Viewer
+
+
+class GraphColoringViewer(Viewer):
+    def __init__(
+        self,
+        num_nodes: int,
+        name: str = "Graph Coloring",
+    ) -> None:
+        self._name = name
+        self._num_nodes = num_nodes
+        colormap_indecies = np.arange(0, 1, 1 / num_nodes)
+        colormap = cm.get_cmap("hsv", num_nodes + 1)
+        self._color_mapping = []
+        for colormap_idx in colormap_indecies:
+            self._color_mapping.append(colormap(colormap_idx))
+        self._color_mapping.append(
+            (0.0, 0.0, 0.0, 1.0)
+        )  # Uncolored node must be black.
+        self._animation: Optional[animation.Animation] = None
+
+    def render(
+        self,
+        state: State,
+        save_path: Optional[str] = None,
+        ax: Optional[plt.Axes] = None,
+    ) -> None:
+        self._clear_display()
+
+        node_scale = 5 + int(np.sqrt(self._num_nodes))
+        node_radius = 0.05 * 5 / node_scale
+
+        if ax is None:
+            fig, ax = plt.subplots(figsize=(node_scale, node_scale))
+            plt.title(f"{self._name}")
+        else:
+            fig = ax.figure
+            ax.clear()
+
+        pos = self._spring_layout(state.adj_matrix)
+
+        for i, (x, y) in enumerate(pos):
+            ax.add_artist(
+                plt.Circle(
+                    (x, y), node_radius, color=self._color_mapping[state.colors[i]]
+                )
+            )
+            ax.text(
+                x, y, str(i), color="white", ha="center", va="center", weight="bold"
+            )
+
+        for i in range(self._num_nodes):
+            for j in range(i + 1, self._num_nodes):
+                if state.adj_matrix[i, j]:
+                    ax.plot(
+                        [pos[i][0], pos[j][0]],
+                        [pos[i][1], pos[j][1]],
+                        color=self._color_mapping[i],
+                        linewidth=0.5,
+                    )
```

Comment:
Maybe this function ca be separated in smaller functions so that it is easier to follow ? 
```python 
    def render(
        self,
        state: State,
        save_path: Optional[str] = None,
        ax: Optional[plt.Axes] = None,
    ) -> None:
        self._clear_display()
        fig, ax = self._get_fig_ax(ax)
        self._render_nodes(ax, state)
        self._render_edges(ax, state)
...
```